File: leet/src/Q_1_10/Q8/Q8.java
package Q_1_10.Q8;

import java.math.BigInteger;

/**
 * @author bockey
 * 字符串变整数，先判断特殊情况，然后取符号，然后在int 范围内操作即可
 *
 *执行用时 :  7 ms  , 在所有 Java 提交中击败了  89.26%  的用户
 *内存消耗 :  36 MB  , 在所有 Java 提交中击败了  86.20%  的用户
 */
public class Q8 {
    public int myAtoi(String str) {
        str = str.trim();
        if (str.startsWith("+-")) {
            return 0;
        }
        if (str.startsWith("+")) {
            str = str.substring(1);
        }
        if (str.isEmpty()) {
            return 0;
        }
        int start = 0;
        int end = str.length();
        if (!(str.charAt(0) == 45 || (str.charAt(0) >= 48 && str.charAt(0) <= 57))) {
            return 0;
        }
        for (int k = start + 1; k < str.length(); k++) {
            if (!(str.charAt(k) >= 48 && str.charAt(k) <= 57)) {
                end = k;
                break;
            }
        }
        String sub = str.substring(start, end);
        if (sub.length() == 1 && sub.equals("-")) {
            return 0;
        }

        try {
            int c = Integer.parseInt(sub);
            return c;
        } catch (Exception e) {
            if (sub.startsWith("-")) {
                return Integer.MIN_VALUE;
            } else {
                return Integer.MAX_VALUE;
            }
        }
//- -45,0-48,9-57
    }

    public static void main(String[] args) {
        Q8 q = new Q8();
        System.out.println(q.myAtoi("+"));
        int c = '+';
        System.out.println("c :" + c);
    }
}

File: leet/src/Q_1_10/Q8/Q8.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

//执行用时 :0 ms, 在所有 Go 提交中击败了100.00%的用户
//内存消耗 :2.3 MB, 在所有 Go 提交中击败了45.45%的用户
func myAtoi(str string) int {
	//trim函数用法，第一个是目标，第二个是首位需要去掉的字符，可以自定义
	str = strings.Trim(str, " ")
	fmt.Println(str)
	if strings.HasPrefix(str, "+-") {
		return 0
	}
	if strings.HasPrefix(str, "+") {
		str = str[1:]
	}
	if len(str) == 0 {
		return 0
	}
	start := 0
	end := len(str)
	if !(str[0] == 45 || (str[0] >= 48 && str[0] <= 57)) {
		return 0
	}
	for k := start + 1; k < len(str); {
		if !(str[k] >= 48 && str[k] <= 57) {
			end = k
			break
		}
		k++
	}
	sub := str[start:end]
	if len(sub) == 1 && sub == "-" {
		return 0
	}
	ints, _ := strconv.Atoi(sub)
	if ints > 2147483647 {
		return 2147483647
	}
	if ints < -2147483648 {
		return -2147483648
	}
	return ints
}

func main() {
	fmt.Println(myAtoi("   -42d "))
}
